src/ignore.h

```diff
@@ -51,6 +52,10 @@ enum {
 	GIT_IGNORE_TRUE = 1,
 };
 
+enum {
+	GIT_IGNORE_TOP_DOWN_DISCOVERY = (1u << 0),
+};
```

Comment:
I'm wondering whether this new flag is needed at all. We are unconditionally setting it for the filesystem iterator and thus we'll always use it for discovering ignores now, right? If so, then I'd say to not introduce the flag but just convert to use `git_vector_rforeach`

Reply:
Right, I was wondering same.  Would it ever be correct to evaluate them bottom up instead of top down?  Seems like no.

Reply:
Thanks for the review and your comments. Using `git_vector_rforeach` was my first choice too. Then I found out that `git_ignore_path_is_ignored` api walks the tree up to discover ignores (https://github.com/libgit2/libgit2/blob/master/src/ignore.c#L327). Please let me know if my understanding is incorrect.

Reply:
Is that correct, though?  I feel like you've discovered that we're walking up and shouldn't be.  It seems like that's true here, too?

Reply:
I think it is fine to walk up in `git_ignore__for_path`. Otherwise, it would not discover `.gitignore` in a parent directory right? 😉

Reply:
Thanks, @ethomson.

From what I can gather, `git_ignore__for_path` is a lower-level function used in a couple of places.  Given a path, it should return `ignores` applicable to that path. For example, when the input path is `a/b/main.c`, it should discover `.gitignore` files in both `a` and `b`.

Since the starting point is `a/b/main.c`, it makes sense to walk up the tree looking for ignores. I think this is the optimal way to do this in this scenario. Because if we start discovering ignore files from `a` here, we would have to put additional checks to traversing down unwanted paths (e.g. imagine other directories such as `a/c`).

One place `git_ignore__for_path` is used is a public API called `git_ignore_path_is_ignored` (https://libgit2.org/libgit2/#HEAD/group/ignore/git_ignore_path_is_ignored). We should not break this functionality.

`filesystem_iterator` on the other hand starts things off from the root (it actually uses `git_ignore__for_path` to initialize the root level, but that largely irrelevant I guess). As it travels down the tree, it loads up `.gitignore` files on its way. It also uses the ignores discovered so far to decide whether to descend or not.

So given the two patterns in use, IMHO, it's valid to have two ways to discover ignore files.

Hopefully, this clears it up. Thanks a lot again.

Reply:
>  I think this is the optimal way to do this in this scenario. Because if we start discovering ignore files from a here, we would have to put additional checks to traversing down unwanted paths (e.g. imagine other directories such as a/c).

Ignore this part 😄 
I think we can change this function to do a top-down lookup.

Reply:
Sorry, I still don't quite get it and I've long given up hope of completely understanding our gitignore logic. At some point in time I feel like we should just rewrite it altogether, I can't imagine git.git's to be this... convoluted.

Anyway. I've just did a quick test to replace `git_vector_foreach` with `git_vector_rforeach` unconditionally and didn't get a single test failure. So either it's fine to _always_ iterate in reverse order indicating we don't need the flag. Or alternatively we're missing a test case for the szenario you're describing.

I hate to pile up more work on your plate, but could you please try to create a testcase for the szenario where your new flag is relevant? I guess that'd help both @ethomson and myself greatly.

Thanks a lot!

Reply:
@pks-t lolz!

It's funny that it didn't fail for you. When I did it, there were a few breaking tests. Let me try that again and perhaps I can give up on completing the tests for `git_path_walk_down` function I just finished writing 😉 .

Reply:
@pks-t you are right! Tests are passing now. Something must have changed in master since attempted this approach. Here's a quick test to show that applying `git_vector_rforeach` is not sufficient.

```
void test_ignore_status__unignore_in_subdir(void)
{
	static const char *test_files[] = {
		"empty_standard_repo/subdir/a.test",
		"empty_standard_repo/a.test",
		NULL
	};

	make_test_data("empty_standard_repo", test_files);
	cl_git_mkfile(
		"empty_standard_repo/.gitignore",
		" a.test\n");

	cl_git_mkfile(
		"empty_standard_repo/subdir/.gitignore",
		"a.test\n"
		"!a.test\n");

	assert_is_ignored("a.test");
	refute_is_ignored("subdir/a.test");
}

```

Could you please give it whirl and let me know. In the meantime, I will continue with my change to this PR.

Thanks a lot again for your support!

Reply:
@pks-t sorry my bad, above tests works. There was an extra space in a rule 😊 
I will make the adjustments to this PR.

Reply:
@pks-t & @ethomson I've made the requested change now. I tried running tests with this patch in a few tags but all of them work. Not sure why tests failed when I first tried it. Sorry about the lengthy dialog. Thanks again!